### PR TITLE
Fix the error case when there are multi top level unions

### PR DESCRIPTION
**Pull request by LakshSingla**
### Description
This is a follow up to the PR #11908. This fixes the bug in top level union all queries when there are more than 2 SQL subqueries are present. For example, if the following query was passed:

```
SELECT channel FROM wikipedia
UNION ALL SELECT channel FROM wikipedia WHERE channel = '#en.wikipedia'
UNION ALL SELECT channel FROM wikipedia WHERE channel = '#de.wikipedia'
```
it would fail to be explained in the newer format, and would revert to the original plan. 
Reason for the bug: In the original PR, I assumed that `DruidUnionRel` can only be the top level DruidRel node after parsing, according to the comment [here](https://github.com/apache/druid/blob/1f95a42bb8b32c314934e8e2ecdeb6569c7e8db3/sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidUnionRel.java#L48), and all the three queries would be children to that top level parent node.

When there are more than 2 queries connected by a top level UNION ALL, then the DruidRel structure looks like this:
```
DruidUnionRel
  DruidRel (query 1)
  DruidUnionRel
    DruidRel (query 2)
    DruidRel (query 3)
```

This PR introduces a recursive method which flattens this structure, and produces the native queries as expected.
<hr>

##### Key changed/added classes in this PR
 * `DruidPlanner`

<hr>

This PR has:
- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

File: sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidPlanner.java [hunk 1]
```diff
@@ -431,25 +431,26 @@ private PlannerResult planExplanation(
...
-    } else {
-      druidQueryList = ImmutableList.of(rel.toDruidQuery(false));
+    try {
+      druidQueryList = flattenOutermostRel(rel)
+          .stream()
+          .map(druidRel -> druidRel.toDruidQuery(false))
+          .collect(Collectors.toList());
+
+    }
+    catch (UnsupportedOperationException unsupportedOperationException) {
```
Comment (abhishekagarwal87) [inline sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidPlanner.java:443]: are you throwing `unsupportedOperationException` somewhere in the code path? I am wondering why is it being caught separately.

File: sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidPlanner.java [hunk 2]
```diff
@@ -431,25 +431,26 @@ private PlannerResult planExplanation(
...
+      log.error("DruidUnionRel can only be the outermost RelNode. This error shouldn't be encountered");
+      throw new ISE("DruidUnionRel is only supported at the outermost RelNode.");
+    }
+    catch (Exception e) {
```
Comment (abhishekagarwal87) [inline sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidPlanner.java:447]: we should log the exception details though.

File: sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidPlanner.java (hunk 1 above)
Comment (LakshSingla) [reply sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidPlanner.java:443]: `DruidUnionRel` throws it for unsupported operations: `https://github.com/apache/druid/blob/1f95a42bb8b32c314934e8e2ecdeb6569c7e8db3/sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidUnionRel.java#L125` (Those operations which can be done on queries which translate directly to native queries). Here `toDruidQuery` can throw the exception.

File: sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidPlanner.java (hunk 2 above)
Comment (LakshSingla) [reply sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidPlanner.java:447]: Updated the details in the unknown exception's case.

File: sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidPlanner.java [hunk 3]
```diff
@@ -431,25 +432,27 @@ private PlannerResult planExplanation(
...
-    } else {
-      druidQueryList = ImmutableList.of(rel.toDruidQuery(false));
+    try {
+      druidQueryList = flattenOutermostRel(rel)
+          .stream()
+          .map(druidRel -> druidRel.toDruidQuery(false))
+          .collect(Collectors.toList());
+
+    }
+    catch (UnsupportedOperationException unsupportedOperationException) {
```
Comment (abhishekagarwal87) [inline sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidPlanner.java:444]: these catch blocks seem unnecessary as we are catching all the exceptions and logging the whole stack trace anyway.

Comment (LakshSingla) [reply sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidPlanner.java]: Updated in the latest commit.

**Review by abhishekagarwal87 — APPROVED** [review]
+1

**Comment by abhishekagarwal87** [comment]
Merged since CI failure is unrelated.

**Pull request merged**